packages/@vue/cli-shared-utils/lib/env.js

```diff
@@ -79,32 +79,33 @@ exports.hasProjectGit = (cwd) => {
 }
 
 let _hasPnpm
-let _hasPnpm3orLater
+let _pnpmVersion
 const _pnpmProjects = new LRU({
   max: 10,
   maxAge: 1000
 })
 
-exports.hasPnpm3OrLater = () => {
-  if (process.env.VUE_CLI_TEST) {
-    return true
-  }
-  if (_hasPnpm3orLater != null) {
-    return _hasPnpm3orLater
-  }
+function getPnpmVersion () {
   try {
-    const pnpmVersion = execSync('pnpm --version', {
+    _pnpmVersion = execSync('pnpm --version', {
       stdio: ['pipe', 'pipe', 'ignore']
     }).toString()
     // there's a critical bug in pnpm 2
     // https://github.com/pnpm/pnpm/issues/1678#issuecomment-469981972
     // so we only support pnpm >= 3.0.0
     _hasPnpm = true
-    _hasPnpm3orLater = semver.gte(pnpmVersion, '3.0.0')
-    return _hasPnpm3orLater
-  } catch (e) {
-    return (_hasPnpm3orLater = false)
+  } catch (e) {}
+  return _pnpmVersion || '0.0.0'
+}
+
+exports.hasPnpmXOrLater = (version) => {
```

Comment:
While a more generic function is appreciated, don't remove the old exports…
Otherwise, you are changing the public API of this package, meaning this commit would be a breaking change.

Reply:
Also, the parameter is a little bit confusing here.
If the function name has `X` in it, I would expect it to accept a version number (`hasPnpmXOrLater(3)`), while it requires a full version string (`hasPnpmXOrLater('3.0.0')`)…

Reply:
Sorry for the inconveniences. I have corrected your concerns. `hasPnpm3OrLater` is now a wrapper for the `hasPnpmVersionOrLater` function with appropriate version number.

Reply:
I don't know why but somehow github doesn't want to add my latest commit ~6h ago into this merge request.